arduino/sketch/sketch_test.go

```diff
@@ -17,158 +17,322 @@ package sketch
+	require.Empty(t, files)
+
+	sketchPath = paths.New("testdata", "SketchPde")
+	files = CheckForPdeFiles(sketchPath)
+	require.Len(t, files, 1)
+	require.Equal(t, sketchPath.Join("SketchPde.pde"), files[0])
+
+	sketchPath = paths.New("testdata", "SketchMultipleMainFiles")
+	files = CheckForPdeFiles(sketchPath)
+	require.Len(t, files, 1)
+	require.Equal(t, sketchPath.Join("SketchMultipleMainFiles.pde"), files[0])
+
+	sketchPath = paths.New("testdata", "SketchSimple", "SketchSimple.ino")
+	files = CheckForPdeFiles(sketchPath)
+	require.Empty(t, files)
+
+	sketchPath = paths.New("testdata", "SketchPde", "SketchPde.pde")
+	files = CheckForPdeFiles(sketchPath)
+	require.Len(t, files, 1)
+	require.Equal(t, sketchPath.Parent().Join("SketchPde.pde"), files[0])
+
+	sketchPath = paths.New("testdata", "SketchMultipleMainFiles", "SketchMultipleMainFiles.ino")
+	files = CheckForPdeFiles(sketchPath)
+	require.Len(t, files, 1)
+	require.Equal(t, sketchPath.Parent().Join("SketchMultipleMainFiles.pde"), files[0])
+
+	sketchPath = paths.New("testdata", "SketchMultipleMainFiles", "SketchMultipleMainFiles.pde")
+	files = CheckForPdeFiles(sketchPath)
+	require.Len(t, files, 1)
+	require.Equal(t, sketchPath.Parent().Join("SketchMultipleMainFiles.pde"), files[0])
+}
+
+func TestNewSketchWithSymlink(t *testing.T) {
+	sketchPath, _ := paths.New("testdata", "SketchWithSymlink").Abs()
+	mainFilePath := sketchPath.Join("SketchWithSymlink.ino")
+	helperFilePath := sketchPath.Join("some_folder", "helper.h")
+	helperFileSymlinkPath := sketchPath.Join("src", "helper.h")
+	srcPath := sketchPath.Join("src")
+
+	// Create a symlink in the Sketch folder
+	os.Symlink(sketchPath.Join("some_folder").String(), srcPath.String())
+	defer srcPath.Remove()
+
+	sketch, err := New(sketchPath)
 	require.NoError(t, err)
-	require.Equal(t, sketchPath.String(), sketch.LocationPath)
-	require.Equal(t, mainFilePath, sketch.MainFile.Path)
-	require.Len(t, sketch.OtherSketchFiles, 0)
-	require.Len(t, sketch.AdditionalFiles, 1)
-	require.Equal(t, "template.ipp", filepath.Base(sketch.AdditionalFiles[0].Path))
-	require.Len(t, sketch.RootFolderFiles, 1)
-	require.Equal(t, "template.ipp", filepath.Base(sketch.RootFolderFiles[0].Path))
+	require.NotNil(t, sketch)
+	require.True(t, sketch.MainFile.EquivalentTo(mainFilePath))
+	require.True(t, sketch.FullPath.EquivalentTo(sketchPath))
+	require.Equal(t, sketch.OtherSketchFiles.Len(), 0)
+	require.Equal(t, sketch.AdditionalFiles.Len(), 2)
+	require.True(t, sketch.AdditionalFiles.Contains(helperFilePath))
+	require.True(t, sketch.AdditionalFiles.Contains(helperFileSymlinkPath))
+	require.Equal(t, sketch.RootFolderFiles.Len(), 0)
+}
+
+func TestNewSketchWithSymlinkLoop(t *testing.T) {
+	sketchPath, _ := paths.New("testdata", "SketchWithSymlinkLoop").Abs()
+	someSymlinkPath := sketchPath.Join("some_folder", "some_symlink")
+
+	// Create a recursive Sketch symlink
+	os.Symlink(sketchPath.String(), someSymlinkPath.String())
```

Comment:
Did you try this test with the original code base? In order to get the hang when compiling a sketch from the command line (I haven't done any in depth investigations into this issue using the Arduino CLI Go code directly), I found two things are necessary:

- There are two symlinks to the parent folder.
- The symlinks are under the `src` subfolder of the sketch.

There is a demo sketch here:
[SymlinkLoopSketch.zip](https://github.com/arduino/arduino-cli/files/6806543/SymlinkLoopSketch.zip)

Arduino CLI no longer hangs when I compile that sketch using the build from this PR, so I do think it's fixed, but I'm just not certain this test covers that problem.

Reply:
It's kinda hard testing this for the original code since we had 3 different ways of creating a Sketch. I'll test the demo Sketch you provided too just to be sure.

Reply:
I wrote a test with your example, works nicely. :woman_dancing: